src/java.base/share/classes/java/lang/IO.java

```diff
@@ -0,0 +1,202 @@
+     */
+    public static void print(Object obj) {
+        System.out.print(obj);
+        System.out.flush();
+    }
+
+    /**
+     * Reads a single line of text from the standard input.
+     * <p>
+     * If necessary, this method first creates an internal
+     * {@link java.nio.charset.CharsetDecoder CharsetDecoder}
+     * to decode the bytes read from the standard input into characters.
+     * It is then wrapped within a {@link java.io.Reader Reader} to
+     * provide character input. These objects are retained for
+     * subsequent use by this method.
+     * <p>
+     * One line is read as if by
+     * {@link java.io.BufferedReader#readLine() BufferedReader.readLine()}
+     * and then the result is returned.
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line-termination characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln() {
+        try {
+            return reader().readLine();
+        } catch (IOException ioe) {
+            throw new IOError(ioe);
+        }
+    }
+
+    /**
+     * Writes a prompt and then reads a line of input.
+     *
+     * <p>
+     * Writes a prompt as if by calling {@code print}, and then reads a single
+     * line of text as if by calling {@link readln readln()}.
+     *
+     * @param prompt the prompt string, may be {@code null}
+     *
+     * @return a string containing the line read from the standard input, not
+     * including any line-termination characters. Returns {@code null} if an
+     * end of stream has been reached without having read any characters.
+     *
+     * @throws IOError if an I/O error occurs
+     */
+    public static String readln(String prompt) {
+        print(prompt);
+        return readln();
+    }
+
+    static BufferedReader br;
+
+    /**
+     * Returns the internal BufferedReader instance used for reading text from
+     * the standard input. The internal BufferedReader is created if necessary
+     * and is cached for future use. Subsequent calls to this method return the
+     * same BufferedReader instance.
+     * <p>
+     * The default charset used when creating the internal BufferedReader is UTF-8.
+     * A different charset maybe specified calling the {@link #setInputEncoding
+     * setInputEncoding} method prior to calling this method or other reading methods
+     * on this class.
+     * <p>
+     * The result of interleaving calls to methods on the internal BufferedReader
+     * (including other methods on this class) with operations on {@code System.in}
+     * is unspecified.
+     *
+     * @return the internal BufferedReader instance
+     */
+    private static synchronized BufferedReader reader() {
+        if (br == null) {
+            br = new BufferedReader(new InputStreamReader(System.in, CHARSET));
```

Comment:
The `print` methods always use `System.out`, so if it was reassigned via `System.setOut`, they would use the new standard output stream. But `System.in` is only read once, meaning `java.lang.IO` won't be affected by `System.setIn`. Is that intended? I guess it could already be covered by "The result of interleaving calls to the `readln` methods with operations on `System.in` is unspecified."

Reply:
Yeah, since System.in needs to be wrapped in an InputStreamReader and a BufferedReader, it doesn't make sense to try to track changes that might occur because of System.setIn.